plugins/debuginfo-install.py

```diff
@@ -79,54 +83,95 @@ def run(self):
         errors_spec = []
 
         for pkgspec in self.opts.package:
-            package_query = sorted(dnf.subject.Subject(pkgspec).get_best_query(
-                self.cli.base.sack).filter(arch__neq='src'), reverse=True)
-            if not package_query:
-                msg = _('No match for argument: %s')
-                logger.info(msg, self.base.output.term.bold(pkgspec))
+            solution = dnf.subject.Subject(pkgspec).get_best_solution(self.base.sack)
+
+            query = solution["query"].filterm(arch__neq='src')
+
+            packages = query.available()._na_dict()
+            # installed versions of packages have priority, replace / add them to the dict
+            packages.update(query.installed()._na_dict())
+
+            # Remove debuginfo packages if their base packages are in the query.
+            # They can get there through globs and they break the installation
+            # of debug packages with the same version as the installed base
+            # packages. If the base package of a debuginfo package is not in
+            # the query, the user specified a debug package on the command
+            # line. We don't want to ignore those, so we will install them.
+            # But, in this case the version will not be matched to the
+            # installed version of the base package, as that would require
+            # another query and is further complicated if the user specifies a
+            # version themselves etc.
+            for na in list(packages.keys()):
+                if na[0].endswith(Package.DEBUGINFO_SUFFIX):
+                    if (na[0][:-len(Package.DEBUGINFO_SUFFIX)], na[1]) in packages:
+                        packages.pop(na)
+                if na[0].endswith(Package.DEBUGSOURCE_SUFFIX):
+                    if (na[0][:-len(Package.DEBUGSOURCE_SUFFIX)], na[1]) in packages:
+                        packages.pop(na)
+
+            if not packages:
+                logger.info(_('No match for argument: %s'), self.base.output.term.bold(pkgspec))
                 errors_spec.append(pkgspec)
-            for pkg in package_query:
-                self._di_install(pkg)
 
-        if not self.dbgdone:
-            logger.info(_("No debuginfo packages available to install"))
+            # attempt to install debuginfo and debugsource for the highest
+            # listed version of the package (in case the package is installed,
+            # only the installed version is listed)
+            for pkg in packages.values():
```

Comment:
I would recommend to add support when 'foo-debuginfo' argument is used. Debuginfo command not only search for packages, but also enebles debuginfo repositories. What about the change like this?
```
            for pkgs in packages.values():
                package = pkgs[0]
                pkg_name = package.name
                if pkg_name.endswith(Package.DEBUGINFO_SUFFIX) or pkg_name.endswith(Package.DEBUGSOURCE_SUFFIX):
                    sltr = dnf.selector.Selector(self.base.sack)
                    sltr.set(pkg=pkgs)
                    self.base.goal.install(select=sltr, optional=not self.base.conf.strict)
                else:
                    if not self._install(package, package.debug_name, solution["nevra"]):
                        self.debuginfo_missing.add(str(package))

                    if not self._install(package, package.debugsource_name, solution["nevra"]):
                        self.debugsource_missing.add(str(package))
```

Reply:
There is a support for ``foo-debuginfo`` arguments. Note that Lukas changed/added the ``debug_name`` and ``debugsource_name`` functions in https://github.com/rpm-software-management/dnf/pull/1425 so that it handles the suffixes better.

Debuginfo repositories are enabled in ``self.configure()``. Or what did you mean?

Reply:
Thanks Pavla I didn't know that. But anyway the proposed change will spare another query, because the search result should be identical.